src/components/App.js

```diff
@@ -0,0 +1,67 @@
+import './App.css'
+import Card from './Card'
+import { useEffect, useState } from 'react'
+import Pagination from './Pagination'
+import getCharacters from '../services/getCharacters'
+
+function App() {
+  const [characters, setCharacters] = useState([])
+  const [cachedCharacters, setCachedCharacters] = useState({})
+  const [totalPages, setTotalPages] = useState(0)
+  const [currentPage, setCurrentPage] = useState(1)
+  const url = 'https://rickandmortyapi.com/api/character?page=' + currentPage
+
+  useEffect(() => {
+    if (!cachedCharacters[currentPage]) {
+      getCharacters(url).then(data => {
+        setCharacters(data.results)
+        setCachedCharacters({
+          ...cachedCharacters,
+          [currentPage]: data.results,
+        })
+        setTotalPages(data.info.pages)
+      })
+    }
+  }, [url])
+
+  function handleClickBack() {
+    setCurrentPage(currentPage - 1)
+    if (cachedCharacters[currentPage - 1]) {
+      setCharacters(cachedCharacters[currentPage - 1])
+    }
+  }
+
+  function handleClickForward() {
+    setCurrentPage(currentPage + 1)
+    if (cachedCharacters[currentPage + 1]) {
+      setCharacters(cachedCharacters[currentPage + 1])
+    }
```

Comment:
I would move this up to the useEffect as the `if` clause. The fetching should then be in the `else`.